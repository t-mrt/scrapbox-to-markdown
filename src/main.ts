import {
  Block,
  CodeBlock,
  Line,
  parse,
  Table,
  Title,
} from "@progfay/scrapbox-parser";

const isTitle = (v: Block): v is Title => {
  return v.type === "title";
};

const isLine = (v: Block): v is Line => {
  return v.type === "line";
};

const isTable = (v: Block): v is Table => {
  return v.type === "table";
};

const isCodeBlock = (v: Block): v is CodeBlock => {
  return v.type === "codeBlock";
};

export const main = (scrapboxAPIText: string): string => {
  console.log(parse(scrapboxAPIText, {}));
  return parse(scrapboxAPIText, {})
    .map((v) => {
      if (isTitle(v)) {
        return v.text && `# ${v.text}`;
      } else if (isLine(v)) {
        const l = v.nodes
          .map((n) => {
            switch (n.type) {
              case "quote":
                break;
              case "helpfeel":
                break;
              case "strongImage":
                break;
              case "strongIcon":
                break;
              case "strong":
                break;
              case "formula":
                break;
              case "decoration":
                break;
              case "code": // `text`
                return n.raw;
              case "commandLine": // $ command
                return `\`${n.raw}\``;
              case "blank":
                break;
              case "image":
                break;
              case "link":
                break;
              case "googleMap":
                break;
              case "icon":
                break;
              case "hashTag":
                break;
              case "numberList":
                break;
              case "plain":
                return n.raw;
              default:
                break;
            }
          })
          .join("");
        if (v.indent === 0) {
          return `${l}`;
        } else {
          return `${"  ".repeat(v.indent - 1) + "- " + l}`;
        }
      } else if (isCodeBlock(v)) {
        return `${v.fileName}`;
      } else if (isTable(v)) {
        return `${v.fileName}`;
      } else {
        console.log(`Unknown type: ${v}`);
      }
    })
    .join("\n");
};

// const PROJECT_NAME = "yux3";
// const PAGE_NAME = "markdownテスト";
// const res = fetch(
//   `https://scrapbox.io/api/pages/${PROJECT_NAME}/${PAGE_NAME}/text`
// )
//   .then((response) => response.text())
//   .then((text) => console.log(main(text)));
