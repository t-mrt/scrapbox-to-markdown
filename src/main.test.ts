import { main } from "./main";

describe("test", () => {
  it("indent", () => {
    const text = `
test
 test
  test
   test
\ttest
\t\ttest
`;
    expect(main(text)).toMatchSnapshot();
  });

  it("title", () => {
    const text = `title-name
body1
body2
`;
    expect(main(text)).toMatchSnapshot();
  });

  it("code", () => {
    const text = `
plain\`code\`plain2
`;
    expect(main(text)).toMatchSnapshot();
  });

  it("code", () => {
    const text = `
$ commandLine
`;
    expect(main(text)).toMatchSnapshot();
  });
});
