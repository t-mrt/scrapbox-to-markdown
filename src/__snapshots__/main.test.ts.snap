// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test code 1`] = `
"
plain\`code\`plain2
"
`;

exports[`test code 2`] = `
"
\`$ commandLine\`
"
`;

exports[`test indent 1`] = `
"
test
- test
  - test
    - test
- test
  - test
"
`;

exports[`test quote 1`] = `
"
> aa
> aaa
> aaa
"
`;

exports[`test title 1`] = `
"# title-name
body1
body2
"
`;
